libqtile/widget/crypto_ticker.py

```diff
@@ -74,14 +82,36 @@ def __init__(self, **config):
         if self.symbol == "":
             self.symbol = "$"
 
+        if self.api in self.QUERY_URL_DICT:
+            self.query_url = self.QUERY_URL_DICT[self.api]
+        else:
+            apis = sorted(self.QUERY_URL_DICT.keys())
+            logger.error(
+                "%s is not a valid API. Use one of the list: %s.",
+                self.api,
+                str(apis),
+            )
+
     @property
     def url(self):
-        return self.QUERY_URL.format(self.crypto, self.currency)
+        return self.query_url.format(self.crypto, self.currency)
 
     def parse(self, body):
+        if self.api not in self.QUERY_URL_DICT:
+            return "Invalid API"
+
         variables = dict()
         variables["crypto"] = self.crypto
         variables["symbol"] = self.symbol
-        variables["amount"] = float(body["data"]["amount"])
+        variables["amount"] = self.getAmountFromBodyResponse(body)
 
         return self.format.format(**variables)
+
+    def getAmountFromBodyResponse(self, body):
+        if (self.api == "binance"):
+            return float(body["price"])
+        
+        if (self.api == "coinbase"):
+            return float(body["data"]["amount"])
+
+        return 0.0
```

Comment:
I guess this 0.0 will be reached in the case where the `api` parameter that was passed from the config wasn't found in `QUERY_URL_DICT`. When that happens it might be best to raise a `ConfigError` (from `libqtile.confreader`) from within the `_configure` method (rather than `__init__`). This causes the widget to be replaced in the bar by a text widget that tells the user that the widget was invalid, and the original widget doesn't get set up. Then we can assume from this method that if `self.api == "binance"` is False, that it _must_ be `"coinbase"`. The `_configure` method would look something like this:
```py
    def _configure(self, qtile, bar):
        if self.api not in self.QUERY_URL_DICT:
            raise ConfigError("Unknown provider passed as 'api' to Cryptoticker")

        GenPollUrl._configure(self, qtile, bar)
        self.query_url = self.QUERY_URL_DICT[self.api]
```
(untested)

Reply:
I made some modifications to your code to adapt it to the new dictionary. Take a look: 25096b15e9602a690d0649e79e195f7ecdce0ccc